pennylane/templates/subroutines/prepselprep.py

```diff
@@ -203,3 +211,34 @@ def target_wires(self):
     def wires(self):
         """All wires involved in the operation."""
         return self.hyperparameters["control"] + self.hyperparameters["target_wires"]
+
+
+def _prod_resources(rep):
+    if rep.op_type == qml.ops.Prod:
+        return {qml.resource_rep(qml.GlobalPhase): 1, **rep.params["resources"]}
+    return {qml.resource_rep(qml.GlobalPhase): 1, rep: 1}
+
+
+def _prepselprep_resources(op_reps, num_control):
+    prod_resources = (_prod_resources(rep) for rep in op_reps)
+    mod_op_reps = tuple(qml.resource_rep(qml.ops.Prod, resources=res) for res in prod_resources)
+    return {
+        qml.resource_rep(qml.Select, op_reps=mod_op_reps, num_control_wires=num_control): 1,
+        qml.resource_rep(qml.StatePrep, num_wires=num_control): 1,
+        qml.resource_rep(
+            qml.ops.Adjoint, base_class=qml.StatePrep, base_params={"num_wires": num_control}
+        ): 1,
+    }
+
+
+# pylint: disable=unused-argument, too-many-arguments
+@qml.register_resources(_prepselprep_resources)
+def _prepselprep_decomp(*_, wires, lcu, coeffs, ops, control, target_wires):
+    coeffs, ops = _get_new_terms(lcu)
```

Comment:
I took a look at this function, and it seems that the introduction of `Prod` is entirely unnecessary. Why not simply accumulate the phase from converting each coefficient into a single `GlobalPhase` gate that we apply separately at the end? There's no reason why we have to end up with a list of `Prod` here, especially when each `Prod` is getting decomposed into the original op and a `GlobalPhase` down the line anyways.

Reply:
Ah I see. That makes sense.